### PR TITLE
Reimplement fetch()

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: repository/src/fetch.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        let mut hit = false;
+        for proof in &finalization_proofs {
+            if csv.verify_last_header_finalization(&proof.proof).is_ok() {
+                last_finalization_proof = proof.clone();
+                last_finalized_commit_hash = *block_commit_hashes.last().unwrap();
+                hit = true;
+            }
+        }
+
+        // If failed, finalize the last-last block.
```
Comment (YoonhoKim0527) [inline repository/src/fetch.rs:201]: I think people can't understand `last-last` meaning. Can you change this comment?

File: repository/src/fetch.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        if !hit && block_commit_hashes.len() > 2 {
```
Comment (YoonhoKim0527) [inline repository/src/fetch.rs:202]: Can we pass the len =1, len = 0 case?

File: repository/src/fetch.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+    // step 3: update finalization
+    this.raw.checkout_clean().await?;
+    this.raw
+        .move_branch(FINALIZED_BRANCH_NAME.into(), last_finalized_commit_hash)
+        .await?;
+    this.raw
+        .move_branch(FP_BRANCH_NAME.into(), last_finalized_commit_hash)
+        .await?;
+    this.raw.checkout(FP_BRANCH_NAME.into()).await?;
```
Comment (YoonhoKim0527) [inline repository/src/fetch.rs:264]: We checkout to FP_BRANCH and finished this method.
I think we have to finish this method with our current working branch. 
-> I think we should re-checkout to the working branch.(like real-git)

File: network/src/dms.rs [hunk 4]
```diff
@@ -667,6 +667,7 @@ mod tests {
             configs,
             Peer {
                 public_key: keys[0].0.clone(),
+                name: format!("{}", keys[0].0),
```
Comment (TomTaehoonKim) [inline network/src/dms.rs:670]: to be updated?

File: repository/src/lib.rs [hunk 5]
```diff
@@ -41,6 +43,12 @@ pub struct Config {
     /// They're added as a remote repo, named `public_#`.
     /// Note that they are not part of the `known_peers`.
     pub mirrors: Vec<String>,
+    /// The distance that if a commit is past this far,
+    /// any forked branch starting from the commit
+    /// will be considered as a long range attack and thus ignored.
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:48]:
```suggestion
    /// The distance that if a commit is past this far,
    /// any forked branch starting from the commit
    /// will be considered a long range attack and thus ignored.
```

File: repository/src/fetch.rs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+async fn add_remotes<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    known_peers: &[Peer],
+) -> Result<(), Error> {
+    for peer in known_peers {
+        let remote_name = peer.name.clone();
+        let remote_url = format!("git://{}/repo", peer.address.ip());
+        if let Err(err) = this.raw.add_remote(remote_name, remote_url.clone()).await {
+            warn!("Failed to add remote({}): {}", remote_url, err);
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:12]:
```suggestion
            warn!("failed to add remote({}): {}", remote_url, err);
```

File: repository/src/fetch.rs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        }
+    }
+    for (i, mirror) in this.config.mirrors.iter().enumerate() {
+        let remote_name = format!("mirror_{}", i);
+        if let Err(err) = this.raw.add_remote(remote_name, mirror.clone()).await {
+            warn!("Failed to add remote({}): {}", mirror, err);
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:18]:
```suggestion
            warn!("failed to add remote({}): {}", mirror, err);
```

File: repository/src/fetch.rs [hunk 8]
```diff
@@ -0,0 +1,269 @@
+use super::*;
+use thiserror::Error;
+
+async fn add_remotes<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    known_peers: &[Peer],
+) -> Result<(), Error> {
+    for peer in known_peers {
+        let remote_name = peer.name.clone();
+        let remote_url = format!("git://{}/repo", peer.address.ip());
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:10]: is this git's configuration or ours?

File: repository/src/fetch.rs [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    for (i, mirror) in this.config.mirrors.iter().enumerate() {
+        let remote_name = format!("mirror_{}", i);
+        if let Err(err) = this.raw.add_remote(remote_name, mirror.clone()).await {
+            warn!("Failed to add remote({}): {}", mirror, err);
+        }
+    }
+    Ok(())
+}
+
+/// Retrieve all the finalization proof
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:24]:
```suggestion
/// Retrieve all the finalization proofs
```

File: repository/src/fetch.rs [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+async fn retrieve_fps<T: RawRepository>(
+    this: &DistributedRepository<T>,
+    remote_branches: &[(String, String, CommitHash)],
+) -> Result<Vec<LastFinalizationProof>, Error> {
+    let mut result = Vec::new();
+    for (remote_name, branch_name, commit_hash) in remote_branches {
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:31]:
```suggestion
        let branch_displayed = format!("{}/{}(at {})", remote_name, branch_name, commit_hash);
```

File: repository/src/fetch.rs [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+async fn retrieve_local_branches<T: RawRepository>(
+    raw: &T,
+) -> Result<HashSet<(Branch, CommitHash)>, Error> {
+    let local_branches = raw.list_branches().await?;
+    let mut result = Vec::new();
+    // TODO: making this concurrent causes a god damn lifetime annoying error
+    for b in local_branches {
+        result.push((b.clone(), raw.locate_branch(b).await?))
+    }
+    Ok(result.into_iter().collect())
```
[the remark this answers is not in the document]
Comment (TomTaehoonKim) [reply repository/src/fetch.rs:61]: > What is this for?

maybe it's what's left when he was trying to make this concurrent
```suggestion
    Ok(result)
```

File: repository/src/fetch.rs [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+            .await?,
+    )?;
+    let last_header = this.get_last_finalized_block_header().await?;
+    let reserved_state = this.get_reserved_state().await?;
+
+    let mut next_step_branches = Vec::new();
+
+    // step 1: update finalization
+    'branch_loop: for (remote_name, branch_name, commit_hash) in remote_branches {
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:125]:
```suggestion
        let branch_displayed = format!("{}/{}(at {})", remote_name, branch_name, commit_hash);
```

File: repository/src/fetch.rs [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+        this.raw
+            .read_semantic_commit(this.raw.locate_branch(FP_BRANCH_NAME.into()).await?)
+            .await?,
+    )?;
+    let last_header = this.get_last_finalized_block_header().await?;
+    let reserved_state = this.get_reserved_state().await?;
+
+    let mut next_step_branches = Vec::new();
+
+    // step 1: update finalization
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:123]:
```suggestion
    // Step 1: update finalization
```

File: repository/src/fetch.rs [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+
+        // Skip if the branch is `fp`.
+        if branch_name == FP_BRANCH_NAME {
+            continue;
+        }
+
+        // Check if the branch is already in the local repo.
+        if local_branches
+            .iter()
+            .any(|(_, commit_hash_)| *commit_hash_ == commit_hash)
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:135]:
```suggestion
            .any(|(_, local_branch_commit_hash)| *local_branch_commit_hash == remote_branch_commit_hash)
```
it doesn't have to be same as this but I think naming should be fixed here

File: repository/src/fetch.rs [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+            Err(e) => return Err(e.into()),
+        };
+
+        // Verify all the incoming commits
+        let mut csv = CommitSequenceVerifier::new(last_header.clone(), reserved_state.clone())
+            .expect("finalized branch is not accepted by CSV");
+        for (new_commit, new_commit_hash) in &commits {
+            if let Err(e) = csv.apply_commit(new_commit) {
+                warn!(
+                    "commit verification failed for branch {}: {} at {}",
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:168]:
```suggestion
                    "commit sequence verification failed for branch {}: {} at {}",
```

File: repository/src/fetch.rs [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+            Err(CommitError::Commit(error, commit)) => {
+                warn!("failed to parse commit {}: {}", commit, error);
+                continue;
+            }
+            Err(e) => return Err(e.into()),
+        };
+
+        // Verify all the incoming commits
+        let mut csv = CommitSequenceVerifier::new(last_header.clone(), reserved_state.clone())
+            .expect("finalized branch is not accepted by CSV");
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:164]:
```suggestion
            .expect("finalized branch is not accepted by Commit Sequence Verifier");
```

File: repository/src/fetch.rs (hunk 1 above)
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:201]:
```suggestion
        // If failed, finalize the second to last block.
```

File: repository/src/fetch.rs [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+        // Store the block commits (including the currently finalized one)
+        let block_commit_hashes = std::iter::once(last_finalized_commit_hash)
+            .chain(
+                commits
+                    .iter()
+                    .filter_map(|(commit, hash)| matches!(commit, Commit::Block(_)).then(|| *hash)),
+            )
+            .collect::<Vec<_>>();
+
+        // Try to finalized the last block with the known finalization proofs.
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:191]:
```suggestion
        // Try to finalize the last block with the known finalization proofs.
```

File: repository/src/fetch.rs [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+        let mut hit = false;
+        for proof in &finalization_proofs {
+            if csv.verify_last_header_finalization(&proof.proof).is_ok() {
+                last_finalization_proof = proof.clone();
+                last_finalized_commit_hash = *block_commit_hashes.last().unwrap();
+                hit = true;
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:197]:
```suggestion
                hit = true;
                break;
```

File: repository/src/fetch.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (YoonhoKim0527) [reply repository/src/fetch.rs:61]: @TomTaehoonKim You mean into_iter casts the type from vector -> hashes?

File: repository/src/fetch.rs (hunk 2 above)
Comment (TomTaehoonKim) [reply repository/src/fetch.rs:202]: In such cases, it means that there is no second to last block commit

File: repository/src/fetch.rs [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+            let last_block = csv.get_block_headers().last().unwrap().clone();
+            last_finalization_proof = LastFinalizationProof {
+                proof: last_block.prev_block_finalization_proof.clone(),
+                height: last_block.height - 1,
+            };
+            last_finalized_commit_hash = block_commit_hashes[block_commit_hashes.len() - 2];
+        }
+    }
+
+    // step 2: create `a` and `b` branches.
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:212]:
```suggestion
    // Step 2: create `a` and `b` branches.
```

File: repository/src/fetch.rs [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+    for (remote_name, branch_name, commit_hash, commit) in next_step_branches {
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:214]:
```suggestion
        let branch_displayed = format!("{}/{}(at {})", remote_name, branch_name, commit_hash);
```

File: repository/src/fetch.rs [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            x => warn!("incomplete remote branch {}: got {:?}", branch_displayed, x),
+        }
+    }
+
+    // step 3: update finalization
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:256]:
```suggestion
    // Step 3: update finalization
```

File: repository/src/fetch.rs [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+                        "b-{:?}",
+                        commit
+                            .to_hash256()
+                            .to_string()
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            x => warn!("incomplete remote branch {}: got {:?}", branch_displayed, x),
```
Comment (TomTaehoonKim) [inline repository/src/fetch.rs:252]:
```suggestion
            c => warn!("incomplete remote branch {}: got {:?}", branch_displayed, c),
```
how about c for a commit

File: repository/src/fetch.rs (hunk 1 above)
Comment (junha1) [reply repository/src/fetch.rs]: fixed.

File: repository/src/fetch.rs (hunk 16 above)
Comment (junha1) [reply repository/src/fetch.rs:164]: `.expect()` message is NOT for the user but for the developers.

Comment (TomTaehoonKim) [reply repository/src/fetch.rs:164]: oh then how is the error handled?

File: repository/src/fetch.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (junha1) [reply repository/src/fetch.rs:61]: fixed

File: repository/src/fetch.rs (hunk 3 above)
Comment (junha1) [reply repository/src/fetch.rs:265]: There is no 'working branch'. Where should I checkout then? Note that the `work` branch belongs to the clone of this repository. (user doesn't directly checkout and add commits to the `work` branch, but just push it)

File: repository/src/fetch.rs (hunk 8 above)
Comment (junha1) [reply repository/src/fetch.rs:10]: You mean `/repo`? That's our protocol. See the next PR.

File: repository/src/fetch.rs (hunk 14 above)
Comment (junha1) [reply repository/src/fetch.rs:135]: fixed

File: repository/src/fetch.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (TomTaehoonKim) [reply repository/src/fetch.rs]: @YoonhoKim0527 I've checked
check and resolve please

File: repository/src/fetch.rs (hunk 2 above)
Comment (TomTaehoonKim) [reply repository/src/fetch.rs:206]: @YoonhoKim0527 check and resolve please

File: repository/src/fetch.rs (hunk 16 above)
Comment (junha1) [reply repository/src/fetch.rs:167]: `.expect()` is not to be 'handled'. It is used if we encounter an error that can't be handled.
More directly saying, if `expect()` hits, that is an obvious bug.

Comment (TomTaehoonKim) [reply repository/src/fetch.rs:167]: understood

**Review by TomTaehoonKim — APPROVED** [review]
Good work

**Review by YoonhoKim0527 — APPROVED** [review]

**Pull request merged**